File: src/main/java/legendarena/api/utils/Rank.java
package legendarena.api.utils;

import org.bukkit.ChatColor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;

/**
 * Rank utilities.
 *
 * @author ThePixelDev
 */
public enum Rank {

    /**
     * Special founder role. Only given to Pixel and Jaden.
     */
    FOUNDER,
    /**
     * Administrator rank.
     */
    ADMIN,
    /**
     * Moderator staff rank.
     */
    MOD,
    /**
     * Helper staff rank.
     */
    HELPER,
    /**
     * "YouTuber" rank.
     */
    YOUTUBE,
    /**
     * "Donor" rank. Only gives cosmetic options.
     */
    MEMBERPLUS,
    /**
     * Default user rank. This is only here as a "catch-all return", really.
     */
    MEMBER;

    /**
     * @param p The player to check
     * @param r The rank to look for
     * @return If the player is ranked after all (true) or not (false)
     */
    public static boolean isRanked(CommandSender p, Rank r) {
        switch(r) {
            case FOUNDER:
                return (p.getName().equals("ThePixelDev") || p.getName().equals("JadenJFilms"));
            case ADMIN:
                return p.hasPermission("legendarena.rank.admin");
            case MOD:
                return p.hasPermission("legendarena.rank.mod");
            case HELPER:
                return p.hasPermission("legendarena.rank.helper");
            case YOUTUBE:
                return p.hasPermission("legendarena.rank.yt");
            case MEMBERPLUS:
                return p.hasPermission("legendarena.rank.memberplus");

            default:
                return false;
        }
    }

    /**
     * @param p The player to check
     * @return The player's rank
     */
    public static Rank getRank(Player p) {
        if(p.getName().equals("ThePixelDev") || p.getName().equals("JadenJFilms")) return FOUNDER;
        else if(p.hasPermission("legendarena.rank.admin")) return ADMIN;
        else if(p.hasPermission("legendarena.rank.mod")) return MOD;
        else if(p.hasPermission("legendarena.rank.helper")) return HELPER;
        else if(p.hasPermission("legendarena.rank.yt")) return YOUTUBE;
        else if(p.hasPermission("legendarena.rank.memeberplus")) return MEMBERPLUS;
        else return MEMBER;
    }

    /**
     * Gets the rank prefix (i.e. colored rank name) for a rank.
     * @param r The rank to get the prefix for
     * @return The rank prefix requested
     */
    public static String getRankPrefix(Rank r) {
        switch(r) {
            case FOUNDER:
                return ChatColor.DARK_RED + "" + ChatColor.BOLD + "FOUNDER";
            case ADMIN:
                return ChatColor.DARK_RED + "ADMIN";
            case MOD:
                return ChatColor.RED + "MOD";
            case HELPER:
                return ChatColor.RED + "HELPER";
            case YOUTUBE:
                return ChatColor.GOLD + "YT";
            case MEMBERPLUS:
                return ChatColor.BLUE + "MEMBER+";

            default:
                return ChatColor.GRAY + "";
        }
    }

    /**
     * Get a pre-built no permissions message.
     * @param r The rank to get the no permissions message of
     * @return The built no permissions message
     */
    public static String noPermissions(Rank r) {
        return ChatUtils.Messages.getCustomMsg("Rank") + "Minimum rank required: " + ChatColor.RED + r;
    }

}

File: src/main/java/legendarena/api/utils/Rank.kt
/*
 * This plugin is copyright Legend Arena, 2014-current. See LICENSE.md for full license file.
 */

package legendarena.api.utils

import legendarena.api.user.User
import org.bukkit.ChatColor
import org.bukkit.command.CommandSender
import org.bukkit.command.ConsoleCommandSender
import org.bukkit.entity.Player

public enum class Rank {

    /**
     * Founder rank. (really just a renamed Owner rank in a sense)
     *
     * Polls [SpecialStaffListUtils] like the Dev rank.
     */
    FOUNDER("", 7),
    /**
     * Developer rank.
     *
     * When the Rank class is asked if the member is a Developer, it polls the [SpecialStaffListUtils] class.
     */
    DEV("", 6),
    /**
     * Administrator rank.
     */
    ADMIN("legendarena.rank.admin", 5),
    /**
     * Moderator rank.
     */
    MOD("legendarena.rank.mod", 4),
    /**
     * Helper rank.
     */
    HELPER("legendarena.rank.helper", 3),
    /**
     * VIP rank.
     *
     * Requirements (any of the following):
     *
     * - 1,500+ subs (YT) / 2,000+ followers (Twitch) and 4+ videos on LA / stream at least three times on LA
     *
     * - Avg. 1k+ views per video / avg. 700+ viewers per stream
     *
     * - Founder/Owner (whichever one you prefer I guess) descretion (overrides the above two)
     */
    VIP("legendarena.rank.vip", 2),
    /**
     * Donor rank. Only gives cosmetic options.
     *
     * Planned to be 10 dollars in the Legend Arena store.
     */
    MEMBERPLUS("legendarena.rank.memberplus", 1),
    /**
     * Default user rank. This is only here as a catch-all, really. Not much else use.
     */
    MEMBER("", 0),
    /**
     * Removes ALL permissions. See the MEMBER rank for the actual default rank.
     */
    NERD("", -1);

    private var permission = ""
    private var internalId = 0

    private constructor() {}

    private constructor(permission: String, internalId: Int) {
        this.permission = permission
        this.internalId = internalId
    }

    public fun isRanked(p: CommandSender): Boolean {
        if(p !is Player) {
            if(p is ConsoleCommandSender)
                return true
            return false
        }

        when(this) {
            Rank.NERD -> {
                var u = User(p)
                if(u.isNoob())
                    return true
                return false
            }
            Rank.MEMBER -> return true
            Rank.DEV -> return SpecialStaffListUtils.isDeveloper(p)
            Rank.FOUNDER -> return SpecialStaffListUtils.isFounder(p)

            else -> return p.hasPermission(permission)
        }
    }

    /**
     * Yes, I realize the function name is Color, not Colour. Fuck it, too lazy to go through more code to fix it.
     */
    public fun getNameColor(): ChatColor {
        when(this) {
            Rank.FOUNDER -> return ChatColor.DARK_AQUA
            Rank.DEV -> return ChatColor.DARK_PURPLE
            Rank.ADMIN -> return ChatColor.DARK_RED
            Rank.MOD -> return ChatColor.RED
            Rank.HELPER -> return ChatColor.GREEN
            Rank.VIP -> return ChatColor.GOLD
            Rank.MEMBERPLUS -> return ChatColor.AQUA

            else -> return ChatColor.GRAY
        }
    }

    public fun getInternalId(): Int {
        return internalId
    }

    public fun isStaff(): Boolean {
        when(this) {
            FOUNDER -> return true
            DEV -> return true
            ADMIN -> return true
            MOD -> return true
            HELPER -> return true

            else -> return false
        }
    }

}